core/commons/che-core-commons-lang/src/main/java/org/eclipse/che/commons/lang/TopologicalSort.java

```diff
@@ -0,0 +1,247 @@
+ * "node". The implementation of the function is completely in the hands of the caller.
+ *
+ * <p>Additionally, a function to extract an "ID" from a node is required. The reasoning behind this
+ * is that it usually is easier to work with identifiers when establishing the predecessors than
+ * with the full node instances. That said, nothing prevents the caller from using the actual node
+ * instance as its ID if the caller so wishes. The consequence of this is that, as a side-effect of
+ * the sorting, the duplicates, as determined by the equality of {@code ID} instances, are removed
+ * from the resulting sorted list.
+ *
+ * @param <N> the type of nodes
+ * @param <ID> the type of a node ID
+ */
+public final class TopologicalSort<N, ID> {
+
+  private final Function<N, ID> identityExtractor;
+  private final Function<N, Set<ID>> directPredecessorsExtractor;
+
+  /**
+   * @param identityExtractor a function to extract some kind of value uniquely identifying a node
+   *     amongst the others.
+   * @param directPredecessorsExtractor a function returning a list of ids of direct predecessors of
+   *     a node
+   */
+  public TopologicalSort(
+      Function<N, ID> identityExtractor, Function<N, Set<ID>> directPredecessorsExtractor) {
+    this.identityExtractor = identityExtractor;
+    this.directPredecessorsExtractor = directPredecessorsExtractor;
+  }
+
+  /**
+   * Given the function for determining the predecessors of the nodes, return the list of the nodes
+   * in topological order. I.e. all predecessors will be placed sooner in the list than their
+   * successors. Note that the input collection is assumed to contain no duplicate entries as
+   * determined by the equality of the {@code ID} type. If such duplicates are present in the input
+   * collection, the output list will only contain the first instance of the duplicates from the
+   * input collection.
+   *
+   * <p>The implemented sort algorithm is stable. If there is no relationship between 2 nodes, they
+   * retain the relative position to each other as they had in the provided collection (e.g. if "a"
+   * preceded "b" in the original collection and there is no relationship between them (as
+   * determined by the predecessor function), the "a" will still precede "b" in the resulting list.
+   * Other nodes may be inserted in between them though in the result).
+   *
+   * <p>The cycles in the graph determined by the predecessor function are ignored and nodes in the
+   * cycle are placed into the output list in the source order.
+   *
+   * @param nodes the collection of nodes
+   * @return the list of nodes sorted in topological order
+   */
+  public List<N> sort(Collection<N> nodes) {
+    // the linked hashmap is important to retain the original order of elements unless required
+    // by the dependencies between nodes
+    LinkedHashMap<ID, NodeInfo<ID, N>> nodeInfos = new LinkedHashMap<>(nodes.size());
```

Comment:
FYI If the intention to precisely predict the size of the map than just size might not be a good predictor. This is the reason why Google Guava has methods for creation of maps with predicted size. But this should not be a problem with non-highload projects and could be treated as premature optimization.

Reply:
In here we know the size **precisely**, though. Notice that for example for the successors I use the default ctor `HashSet<>()` because I have no way of predicting the target size.

IMHO, trying to "guess" an initial/needed size of the collections is not a premature optimization but just a good practice. In https://github.com/rhq-project/rhq we were able to reduce the memory consumption by 30% (!!!) by just thinking about and limiting the initial capacity of the collections involved.

Reply:
In my opinion, usage of a precise size of another map is not reliable to set capacity of the map. The reason that capacity != size. With the default load factor 0.75 capacity is 25% bigger. See https://stackoverflow.com/questions/30220820/difference-between-new-hashmapint-and-guava-maps-newhashmapwithexpectedsizein

If I'm missing some important point why this work otherwise in your case. And would be happy to understand this maps sizing issue better.

Reply:
Ok, seems like you got me here :)

Reply:
I actually thought that the behavior was the same for `*List(int capacity)` as well as `*Set` and `*Map` cases. Thanks for pointing out the difference to me.

Reply:
I didn't mean to get you. This is a tricky part of java that we discussed a lot back when I was in the platform team, so now we can share this knowledge.